File: lib/otel_sample/otel_sample_activity.ex
defmodule OtelSample.Activity do
  use TemporalSdk.Activity
  @moduledoc false
  require OpenTelemetry.Tracer, as: Tracer

  @impl true
  def execute(_context, _input) do
    baggage = :otel_baggage.get_all()
    baggage_attr = Enum.map(baggage, fn {k, {v, _meta}} -> {k, v} end)
    Tracer.set_attributes(baggage_attr)

    Tracer.with_span "my_activity_span", %{:attributes => baggage_attr} do
      Process.sleep(20)
    end

    ["ok"]
  end
end

File: src/otel_sample/otel_sample_activity.erl
-module(otel_sample_activity).

% elp:ignore W0012 W0040
-moduledoc false.

-export([execute/2]).

-include_lib("temporal_sdk/include/activity.hrl").
-include_lib("opentelemetry_api/include/otel_tracer.hrl").

execute(_Context, _Input) ->
    Baggage = otel_baggage:get_all(),
    BaggageAttr = maps:map(fun(_K, {V, _Meta}) -> V end, Baggage),
    ?set_attributes(BaggageAttr),
    ?with_span(~"my_activity_span", #{attributes => BaggageAttr}, fun(_SpanCtx) ->
        timer:sleep(20)
    end),
    [~"ok"].
